final-project/application.py

```diff
@@ -25,103 +25,177 @@ def after_request(response):
         response.headers["Pragma"] = "no-cache"
         return response
 
-# configure session to use filesystem (instead of signed cookies)
+# Configure session to use filesystem (instead of signed cookies).
 app.config["SESSION_FILE_DIR"] = mkdtemp()
 app.config["SESSION_PERMANENT"] = False
 app.config["SESSION_TYPE"] = "filesystem"
 Session(app)
 
-# configure CS50 Library to use SQLite database
+# Configure CS50 Library to use SQLite database.
 db = SQL("sqlite:///final.db")
 
-# users table
-# CREATE TABLE users (id INTEGER PRIMARY KEY NOT NULL, username TEXT UNIQUE NOT NULL, email TEXT UNIQUE NOT NULL, hash TEXT NOT NULL, first TEXT, last TEXT);
 
 @app.route("/")
 @login_required
 def index():
+    """Homepage"""
 
-    # keep user logged in
+    # Keep the user logged in
     session.get("user_id")
 
+    # TODO
     return apology("TODO")
 
 
 @app.route("/categories")
 @login_required
 def categories():
+    """Page of categories to choose from"""
 
     # keep user logged in
     session.get("user_id")
 
+    # TODO
     return apology("TODO")
 
 @app.route("/product")
 # @login_required
 def product():
+    """View an individual product"""
 
-    # keep user logged in
-    session.get("user_id")
-
-    # get name of product
-    product = "%" + request.args.get("product") + "%"
+    # Keep the user logged in.
+    user_id = session.get("user_id")
 
-    try:
-        #Get product from the database
-        product_info = db.execute("SELECT * FROM products WHERE product_name LIKE :product", product = product)
-        # Check whether products were found
-        product_check = product_info[0]
+    # Request product name from user.
+    def request_product():
```

Comment:
Why do you have nested functions? Do you need closures? This increases complexity.

Reply:
It does seem like it increases complexity. How should I write single-purpose methods then? I did not find any good explanations online. All they seemed to do was reference the idea of abstraction.

If you have a particular lesson in mind, related to single-purpose methods, please tell me. It will help me narrow down what to look for. This was my attempt to address your earlier comment.

Reply:
You can just have functions live at the same level as other functions. There's no need to nest them.